### PR TITLE
Add `--incompatible_no_implicit_watch_label`

### PR DESCRIPTION
**Pull request by fmeum**
When the flag is flipped, `ctx.path` no longer watches paths resolved from `Label` parameters and other methods on `ctx` no longer watch such paths even with `watch = "no"`.

### CODE REVIEW

**Comment by iancha1992** [comment]
@fmeum should we include this in `release-7.4.0`?

**Comment by fmeum** [comment]
@bazel-io fork 7.4.0

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java [hunk 1]
```diff
@@ -1352,21 +1352,17 @@ during the analysis phase and thus cannot depends on a target result (the \
...
-    if (path instanceof String) {
-      return new StarlarkPath(this, workingDirectory.getRelative(path.toString()));
-    } else if (path instanceof Label label) {
-      return getPathFromLabel(label);
-    } else if (path instanceof StarlarkPath starlarkPath) {
-      return starlarkPath;
-    } else {
+  protected StarlarkPath getPath(Object path) throws EvalException, InterruptedException {
+    return switch (path) {
+      case String ignored -> new StarlarkPath(this, workingDirectory.getRelative(path.toString()));
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java:1360]:
```suggestion
      case String s -> new StarlarkPath(this, workingDirectory.getRelative(s));
```

less surprising this way, I think ("why the ignore?")

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java [hunk 2]
```diff
@@ -1352,21 +1352,17 @@ during the analysis phase and thus cannot depends on a target result (the \
                     + " a path from.")
       })
   public StarlarkPath path(Object path) throws EvalException, InterruptedException {
-    return getPath("path()", path);
+    return getPath(path);
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java:1355]: probably just merge these two methods. if we ever need to un-merge, the starlark version can be renamed to 'pathForStarlark' or something.

**Review by Wyverald — COMMENTED** [review]
maybe add a shell integration test, with the flag on and off?

**Comment by meteorcloudy** [comment]
Can you file an incompatible change issue to document this flag?
https://github.com/bazelbuild/bazel/issues/new/choose

**Review by Wyverald — APPROVED** [review]

**Comment by fmeum** [comment]
@meteorcloudy I created https://github.com/bazelbuild/bazel/issues/23861.

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!

**Pull request closed without merging**